modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java

```diff
@@ -3822,26 +3823,34 @@ public IgniteInternalFuture<Boolean> dynamicStartCachesByStoredConf(
 
             Iterator<byte[]> grpKeysIter = F.isEmpty(grpKeys) ? null : grpKeys.iterator();
 
+            HashMap<String, byte[]> grpKeyMap = new HashMap<>();
+
             for (StoredCacheData ccfg : storedCacheDataList) {
-                assert grpKeysIter == null || !ccfg.config().isEncryptionEnabled() || grpKeysIter.hasNext();
+                CacheConfiguration<?, ?> cacxheCfg = ccfg.config();
 
-                byte[] encCacheKey = grpKeysIter != null && ccfg.config().isEncryptionEnabled() ? grpKeysIter.next() : null;
+                byte[] encCacheKey = grpKeyMap.get(cacxheCfg.getGroupName());
```

Comment:
It seems to me that using `getGroupName` here might be a problem. 
Let's assume the user starts two new caches without specifying a cache group. It means that two different cache groups should be created, right? On the other hand, after your change, we will use the same encryption key for both.

WDYT? Perhaps I am missing something?

Reply:
You are right because in the case when the cache is self-grouped, we have a null in the configuration for the group name. I fixed it. using acache name instead of a group.
A test of this case was also added.